src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java

```diff
@@ -182,7 +197,15 @@ private ModifiableRealmIdentity getRealmIdentity(final Principal principal, fina
 
     private DirContext obtainContext() throws RealmUnavailableException {
         try {
-            return dirContextSupplier.get();
+            synchronized (this.listenersPendingRegistration) {
+                DirContext ctx = dirContextSupplier.get();
```

Comment:
As context obtaining is network operation often (which can be time consuming everytime), it would be better to have it outside of the synchronized section.